lib/query-builder.js

```diff
@@ -142,51 +180,69 @@ class QueryBuilder {
         // Some nodes won't have a schema representation. For example, __typename.
         if (selectionSchemaNode) {
           if (!field.subgraphs.has(this.subgraph)) {
-            let existing = ctx.followups.get(node)
+            let followup = ctx.followups.get(node)
+
+            if (!followup) {
+              const typeName = type.schemaNode.name
+
+              const entity = this.subgraph.entities[typeName]
+              if (!entity) {
+                const keys = transitiveKeys(typeName, this.subgraph.entities)
+                if (!keys) {
+                  // TODO onError
+                  throw new Error(`Unable to resolve entity ${typeName} in subgraph ${this.subgraph.name}`)
+                }
+                keyFields = keys
+              } else {
+                keyFields = entity.keys
+              }
 
-            if (existing === undefined) {
               // TODO(cjihrig): Double check the next line.
               const nextSubgraph = Array.from(field.subgraphs)[0]
-              const typeName = type.schemaNode.name
               // TODO(cjihrig): Throw if this doesn't exist.
-              const entity = nextSubgraph.entities[typeName]
-              const {
-                adapter,
-                foreignKeyFields,
-                primaryKeyFields,
-                referenceListResolverName: resolverName
-              } = entity
-
-              keyFields = primaryKeyFields ?? foreignKeyFields
-
-              existing = {
-                node,
-                schemaNode,
-                path: ctx.path.slice(),
-                type,
-                fields: [],
-                resolverName,
-                root: false,
-                subgraph: nextSubgraph,
-                info: this.info,
-                adapter,
-                types: this.types
+              const entityOnNextSubgraph = nextSubgraph.entities[typeName]
+              const { referenceListResolverName: resolverName, argsAdapter } = entityOnNextSubgraph
+
+              if (resolverName) {
+                followup = createFollowup({
+                  ctx,
+                  node,
+                  schemaNode,
+                  type,
+                  resolverName,
+                  argsAdapter,
+                  info: this.info,
+                  types: this.types,
+                  subgraph: nextSubgraph,
+                  keys: entityOnNextSubgraph.keys.filter(k => k.type === typeName),
+                  originKeys: keyFields.filter(k => k.type === typeName)
```

Comment:
What if there is more than one field with the same type?

Reply:
as for now, nothing

> current limitations
> entities keys can be of one field, no composite keys